Formula/redo.rb

```diff
@@ -0,0 +1,38 @@
+class Redo < Formula
+  include Language::Python::Virtualenv
+
+  desc "Implements djb's redo: an alternative to make"
+  homepage "https://redo.rtfd.io/"
+  url "https://github.com/apenwarr/redo/archive/redo-0.42a.tar.gz"
+  sha256 "0ff0fd29f2767fe33374cfac97fd13b3dd27207bdacbeb5e02039b992d4fb815"
+
+  depends_on "python"
+
+  resource "Markdown" do
+    url "https://files.pythonhosted.org/packages/98/79/ce6984767cb9478e6818bd0994283db55c423d733cc62a88a3ffb8581e11/Markdown-3.2.1.tar.gz"
+    sha256 "90fee683eeabe1a92e149f7ba74e5ccdc81cd397bd6c516d93a8da0ef90b6902"
+  end
+
+  resource "beautifulsoup4" do
+    url "https://files.pythonhosted.org/packages/3b/e4/7cfc641f11e0eef60123912611a5c9ee7d4638da7325878b695b9ae4bb6f/beautifulsoup4-4.9.0.tar.gz"
+    sha256 "594ca51a10d2b3443cbac41214e12dbb2a1cd57e1a7344659849e2e20ba6a8d8"
+  end
+
+  def install
+    venv = virtualenv_create(libexec, "python3")
+    venv.pip_install resources
+    # Set the interpreter so that ./do install can find the pip installed
+    # resources
+    open("redo/whichpython", "w") do |file|
+      file.puts "#{libexec}/bin/python3"
+    end
+    ENV["DESTDIR"] = ""
+    ENV["PREFIX"] = prefix
+    system "./do", "install"
+    man1.install Dir["docs/*.1"]
+  end
+
+  test do
+    system "#{bin}/redo", "--version"
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
To preface this, the current submission was (for the most part) lazily copied over from the [previous `redo` formula](https://github.com/Homebrew/homebrew-core/commit/73c167ea6fc15830263758389761abafbe592b62) (before it was deleted), so I appreciate the feedback.

Anyway, I took a look at the [README.md](https://github.com/apenwarr/redo/README.md), and it implies that testing occurs _before_ installation. I tried inserting the line
```ruby
system "./do", "-j10", "test"
```
before the installation command, but the docs failed to build because something set the `CC` environment variable behind its back.

<details>
<summary>Output of `./do -j10 test`</summary>

```
==> ./do -j10 test
do  bin/all
do    redo/version/all
do      redo/version/vars
do        redo/version/gitvars
do      redo/version/_version.py
do    redo/py
do    redo/sh
 dash...               missing
 /usr/xpg4/bin/sh...   missing
 ash...                missing
 posh...               missing
 lksh...               missing
 mksh...               missing
 ksh...                failed    F12 F24 s47 s47a F47b W89c F102 W118
 ksh88...              missing
 ksh93...              missing
 pdksh...              missing
 zsh...                warnings  W2a W47c W89c W118
 bash...               warnings  W5 W32 W47c s60 W118
 busybox...            missing
 /bin/sh...            warnings  W5 W32 W47c s60 W94 W118
Selected mostly good shell: /bin/zsh
do    bin/list
do    bin/redo-always
do    bin/redo-ifchange
do    bin/redo-ifcreate
do    bin/redo-log
do    bin/redo-ood
do    bin/redo
do    bin/redo-sources
do    bin/redo-stamp
do    bin/redo-targets
do    bin/redo-unlocked
do    bin/redo-whichdo
do: Removing stamp files...
redo  bin/all
redo    redo/version/all
do: Incremental mode. Use -c for clean rebuild.
do  test
do    bin/all
do      redo/version/all
do    minimal/test

Testing "_starts_ends_with" in ./do.test:
! _starts_ends_with:1  '0' = retval true  ok
! _starts_ends_with:2  '1' = retval false  ok
! _starts_ends_with:3  '0' = retval _startswith x.test x.  ok
! _starts_ends_with:4  '1' = retval _startswith x.test y.  ok
! _starts_ends_with:5  '1' = retval _startswith     ok
! _starts_ends_with:6  '1' = retval _startswith x.test x*  ok
! _starts_ends_with:7  '1' = retval _startswith do.test do*  ok
! _starts_ends_with:8  '0' = retval _endswith x.test .test  ok
! _starts_ends_with:9  '1' = retval _endswith x.test .best  ok
! _starts_ends_with:10  '1' = retval _endswith     ok
! _starts_ends_with:11  '1' = retval _endswith x.test *.test  ok
! _starts_ends_with:12  '1' = retval _endswith do.test *.test  ok

Testing "_dirsplit" in ./do.test:
! _dirsplit:1  '/a/b/' = /a/b/  ok
! _dirsplit:2  'c' = c  ok
! _dirsplit:3  '/a/b/c/' = /a/b/c/  ok
! _dirsplit:4  '' =   ok
! _dirsplit:5  '/' = /  ok
! _dirsplit:6  '' =   ok

Testing "_relpath" in ./do.test:
! _relpath:1  'a/b/c' = _relpath /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x/a/b/c /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _relpath:2  '../a/b/c' = _relpath /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x/../a/b/c /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _relpath:3  '' = _relpath /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _relpath:4  'a/b/c' = _relpath a/b/c /  ok
! _relpath:5  'a/b/c' = _relpath /a/b/c /  ok
! _relpath:6  '' = _relpath / /  ok
! _relpath:7  '../lib' = _relpath /usr/lib /usr/bin  ok
! _relpath:8  '../lib/' = _relpath /usr/lib/ /usr/bin  ok
! _relpath:9  '../..' = _relpath / /usr/bin  ok
! _relpath:10  '..' = _relpath .. /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/fakedir  ok
! _relpath:11  '../' = _relpath ../ /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/fakedir  ok
! _relpath:12  '../fakedir' = _relpath ../fakedir /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/fakedir  ok

Testing "_normpath" in ./do.test:
! _normpath:1  '/usr/lib' = _normpath /usr/../usr/bin/../lib /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _normpath:2  '/' = _normpath /a/b/c/../../.. /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _normpath:3  '/' = _normpath / /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _normpath:4  '../a' = _normpath ../a /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _normpath:5  '../a/b' = _normpath ../a/b /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x/fakedir  ok
! _normpath:6  '..' = _normpath .. /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x/fakedir  ok
! _normpath:7  'tuv' = _normpath a/b/../../tuv /  ok
! _normpath:8  '' = _normpath a/b/../.. /  ok
! _normpath:9  '..' = _normpath ../ /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _normpath:10  '..' = _normpath .. /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok

Testing "_realpath" in ./do.test:
! _realpath:1  '/usr/__does_not/b' = _realpath /usr/__does_not/a/../b /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/x  ok
! _realpath:2  'foo' = _realpath y/x/x/x/x/x/../foo /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal  ok
! _realpath:3  '/private/tmp/redo-20200420-91111-cwbrv9/' = _realpath /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/y/x/x/x/x/x/../foo /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal  ok
! _realpath:4  'foo/blam' = _realpath y/x/x/x/x/x/../foo/spam/../blam /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal  ok
! _realpath:5  '/private/tmp/redo-20200420-91111-cwbrv9/' = _realpath /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal/y/x/x/../foo/spam/../blam /private/tmp/redo-20200420-91111-cwbrv9/redo-redo-0.42a/minimal  ok

Testing "_find_dofile" in ./do.test:
! _find_dofile:1  'test.do' = _find_dofiles test  ok
! _find_dofile:2  'test.do' = _find_dofile test  ok
! _find_dofile:3  'x.y.zz.do default.y.zz.do default.zz.do' = _find_dofiles x.y.zz  ok
! _find_dofile:4  'x.y.zz.do default.y.zz.do default.zz.do ' = _find_dofiles x.y.zz  ok
! _find_dofile:4  '../default.zz.do' = _find_dofile x.y.zz  ok
! _find_dofile:4  '1' = 1  ok
! _find_dofile:5  'x.y.z.do default.y.z.do default.z.do def' = _find_dofiles x.y.z  ok
do    docs/test
do      docs/cookbook/all
do        docs/cookbook/c/all
do          docs/cookbook/c/all.each
do            docs/cookbook/c/allconfig
do              docs/cookbook/c/arches
../configure: Fatal: 'CC' environment variable is set.
  This can cause inconsistent builds.
  Pass variables as arguments to ../configure instead.
  Example: ../configure CC='clang'
do:           allconfig: got exit code 4
do:         all.each: got exit code 1
do:       all: got exit code 1
do:     all: got exit code 1
do:   test: got exit code 1
do: test: got exit code 1
./do: minimal/do test failed
```
</details>

<details>
<summary>Configuration</summary>

```
HOMEBREW_VERSION: 2.2.13-102-ga5b6c5f
ORIGIN: https://github.com/Homebrew/brew
HEAD: a5b6c5f2621298c879aa205754ae01d5417572d9
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 7c1e7551cd6b1a71c7fa425b066e55a575e40cf8
Core tap last commit: 24 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.yF2c7etjqX/org.macosforge.xquartz:0
HOMEBREW_EDITOR: /usr/local/bin/cot
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 4
CPU: quad-core 64-bit broadwell
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby
Clang: 11.0 build 1100
Git: 2.26.1 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 14, 1.8.0_121
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: 11.3.1
XQuartz: 2.7.11 => /opt/X11
```
</details>

<details>
<summary>ENV</summary>

```
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
MAKEFLAGS: -j4
CMAKE_PREFIX_PATH: /usr/local/opt/openssl@1.1:/usr/local/opt/readline:/usr/local/opt/sqlite:/usr/local
CMAKE_INCLUDE_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/libxml2:/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
CMAKE_FRAMEWORK_PATH: /usr/local/opt/python/Frameworks
PKG_CONFIG_PATH: /usr/local/opt/openssl@1.1/lib/pkgconfig:/usr/local/opt/readline/lib/pkgconfig:/usr/local/opt/sqlite/lib/pkgconfig:/usr/local/opt/xz/lib/pkgconfig:/usr/local/opt/python/lib/pkgconfig
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/10.14
HOMEBREW_GIT: git
HOMEBREW_SDKROOT: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk
ACLOCAL_PATH: /usr/local/share/aclocal
PATH: /usr/local/Homebrew/Library/Homebrew/shims/mac/super:/usr/local/opt/gdbm/bin:/usr/local/opt/openssl@1.1/bin:/usr/local/opt/sqlite/bin:/usr/local/opt/xz/bin:/usr/local/opt/python/bin:/usr/bin:/bin:/usr/sbin:/sbin
```
</details>

Any suggestions on how to fix this? I'm sure I'm missing something obvious here.

Reply:
The tests from upstream are usually a little different. For homebrew we just want to check 1 common command for a user to see if we packaged it correctly.

Reply:
A test has been written based on [the "Hello, world!" example given in the docs](https://redo.readthedocs.io/en/latest/cookbook/hello/). It's certainly not exhaustive, a and I want to know if you want anything more comprehensive.